_posts/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1.md

```diff
@@ -0,0 +1,117 @@
+
+The next part will dig into a major break with the original mechanism, namely circumventing the alignment phase — hence the name unaligned checkpoints. Stay tuned for the second part, which explains how unaligned checkpoints work and how they guarantee consistent checkpoint times under backpressure.  
+
+
+## Appendix 1 - On Backpressure
+
+Backpressure refers to the behavior where a slow receiver (e.g. of data/requests) makes the senders slow down in order to not overwhelm the receiver, something that can result in possibly dropping some of the processed data or requests. This is a crucial and very much desirable behavior for systems where completeness/correctness is important. Backpressure is implicitly implemented in many of the most basic building blocks of distributed communication, such as TCP Flow Control, bounded (blocking) I/O queues, poll-based consumers, etc.
+
+Apache Flink implements back-pressure across the entire data flow graph. A sink that (temporarily) cannot keep up with the data rate will result in the source connectors slowing down and pulling data out of the source systems more slowly. We believe that this is a good and desirable behavior, because backpressure is not only necessary in order to avoid overwhelming the memory of a receiver (thread), but can also prevent different stages of the streaming application from drifting apart too far.
+
+Consider the example below:
+- We have a source (let’s say reading data from Apache Kafka), parsing data, grouping and aggregating data by a key, and writing it to a sink system (some database).
+- The application needs to re-group data by key between the parsing and the grouping/aggregation step.
+
+<span>
+	<center>
+	<img vspace="8" style="width:75%" src="{{site.baseurl}}/img/blog/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1/from-aligned-to-unaligned-checkpoints-part-1-imge-5.png" />
+	</center>
+</span>
+
+Let’s assume we use a __non-backpressure approach__, like writing the data to a log/MQ for the data re-grouping over the network (cf. [Kafka Streams](https://docs.confluent.io/current/streams/architecture.html#backpressure)). If the sink is now slower than the remaining parts of the streaming application (which can easily happen), the first stage (source and parse) will still work as fast as possible to pull data out of the source, parse it, and put it into the log for the shuffle. That intermediate log will accumulate a lot of data, meaning it needs significant capacity, so that in a worst case scenario can hold a full copy of the input data or otherwise result in data loss (when the drift is greater than the retention time).
```

Comment:
```suggestion
Let’s assume we use a __non-backpressure approach__, like writing the data to a log/MQ for the data re-grouping over the network (the approach used by [Kafka Streams](https://docs.confluent.io/current/streams/architecture.html#backpressure)). If the sink is now slower than the remaining parts of the streaming application (which can easily happen), the first stage (source and parse) will still work as fast as possible to pull data out of the source, parse it, and put it into the log for the shuffle. That intermediate log will accumulate a lot of data, meaning it needs significant capacity, so that in a worst case scenario can hold a full copy of the input data or otherwise result in data loss (when the drift is greater than the retention time).
```

Reply:
Had to look up what cf. means, it's not as commonly used as _i.e._ or _e.g._.